source/extensions/quic_listeners/quiche/platform/quic_port_utils_impl.cc

```diff
@@ -0,0 +1,47 @@
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include "extensions/quic_listeners/quiche/platform/quic_port_utils_impl.h"
+
+#include "envoy/network/address.h"
+
+#include "common/common/assert.h"
+#include "common/network/utility.h"
+
+#include "test/test_common/environment.h"
+#include "test/test_common/network_utility.h"
+
+namespace quic {
+
+int QuicPickUnusedPortOrDieImpl() {
+  std::vector<Envoy::Network::Address::IpVersion> supported_versions =
+      Envoy::TestEnvironment::getIpVersionsForTest();
+  ASSERT(!supported_versions.empty());
+  // Checking availability under corresponding supported version if test
+  // supports v4 only or v6 only.
+  // If it supports both v4 and v6, checking availability under v6 with IPV6_V6ONLY
+  // set to false is sufficient because such socket can be used on v4-mapped
+  // v6 address.
+  Envoy::Network::Address::IpVersion ip_version = supported_versions.size() == 1
+                                                      ? supported_versions[0]
+                                                      : Envoy::Network::Address::IpVersion::v6;
+  // Only try to get a port for limited times.
+  for (size_t i = 0; i < 300; ++i) {
+    auto addr_port = Envoy::Network::Utility::parseInternetAddressAndPort(
+        fmt::format("{}:{}", Envoy::Network::Test::getAnyAddressUrlString(ip_version), /*port*/ 0),
+        /*v6only*/ false);
+    ASSERT(addr_port != nullptr);
+    addr_port = Envoy::Network::Test::findOrCheckFreePort(
+        addr_port, Envoy::Network::Address::SocketType::Datagram);
+    if (addr_port != nullptr && addr_port->ip() != nullptr) {
+      // Find a port.
+      return addr_port->ip()->port();
+    }
```

Comment:
@curiouserrandy has also written code like this, so might have comments.

Reply:
This looks like a lot of hassle to do something the OS will do for you (i.e. pick a port if you bind port 0), and additionally racy as well.  Can you give me a sense as to why you're doing it this way as opposed to using bind(..., 0) at point of use?

I also share Harvey's concern about mixing apparently production and apparently test code.

Reply:
see my reply to Harvey's comment

Reply:
Ah, gotcha.  Pity :-}.  Next question: As I read the code (including findOrCheckFreePort) what is being done is that the address passed is being bound with port == 0, the port returned by the kernel is being returned, and the bound socket is being released.  Why is that wrapped in a "try 300 times" loop?  If a bind with port 0 fails, I'd think there's something badly wrong somewhere, and retrying wouldn't help.  

(Meta-comment: I'm still in the "asking questions to understand the code" phase of the review, so if you and Harvey get to closure, don't block on me.)

Reply:
> Ah, gotcha. Pity :-}. Next question: As I read the code (including findOrCheckFreePort) what is being done is that the address passed is being bound with port == 0, the port returned by the kernel is being returned, and the bound socket is being released. Why is that wrapped in a "try 300 times" loop? If a bind with port 0 fails, I'd think there's something badly wrong somewhere, and retrying wouldn't help.
Would temporary running out of port lead to bind failure on port 0? Or some other transient system failure?

Reply:
Yeah, I think the loop made sense when we were doing v4/v6 checks but now that we've done away with that I suspect we can kill the loop off entirely.

It could be transient but it's functional equivalent to sleep(1)-and-try-again so I think we can skip

Reply:
done